### PR TITLE
Support for N-ary operations and common intrinsics (towards 339)

### PR DESCRIPTION
**Pull request by arporter**
This is my second attempt (was previously #417).

### CODE REVIEW

**Comment by arporter** [comment]
I think is this is now ready for first review from either @sergisiso or @rupertford.

**Comment by sergisiso** [comment]
@arporter  Overall the PR looks good, I just made a few in-line comments. Tests pass but travis hasn't generated a coverage report, if there isn't one after addressing the current comments I will check manually.

File: src/psyclone/psyGen.py [hunk 1]
```diff
@@ -119,6 +115,7 @@ def colored(text, _):
                        "Reference": "yellow",
                        "BinaryOperation": "blue",
                        "UnaryOperation": "blue",
+                       "NaryOperation": "blue",
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: Oh yes, that makes more sense.

File: src/psyclone/psyGen.py [hunk 2]
```diff
@@ -6080,38 +6095,23 @@ def function_format(function_str, expr_str):
...
 
-    :param operator: Enumerated type capturing the binary operator.
-    :type operator: :py:class:`psyclone.psyGen.BinaryOperation.Operator`
-    :param parent: the parent node of this BinaryOperation in the PSyIR.
-    :type parent: :py:class:`psyclone.psyGen.Node`
     '''
     Operator = Enum('Operator', [
         # Arithmetic Operators
-        'ADD', 'SUB', 'MUL', 'DIV', 'REM', 'POW',
+        'ADD', 'SUB', 'MUL', 'DIV', 'REM', 'POW', 'MODULUS', 'SUM',
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: I didn't even know about MODULO. Changed back to REM!

File: src/psyclone/psyGen.py [hunk 3]
```diff
@@ -6410,7 +6407,21 @@ class Fparser2ASTProcessor(object):
     unary_operators = OrderedDict([
         ('+', UnaryOperation.Operator.PLUS),
         ('-', UnaryOperation.Operator.MINUS),
-        ('.not.', UnaryOperation.Operator.NOT)])
+        ('.not.', UnaryOperation.Operator.NOT),
+        ("abs", UnaryOperation.Operator.ABS),
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: Fair point - changed.

File: src/psyclone/psyGen.py [hunk 4]
```diff
@@ -6431,7 +6442,17 @@ class Fparser2ASTProcessor(object):
         ('>', BinaryOperation.Operator.GT),
         ('.gt.', BinaryOperation.Operator.GT),
         ('.and.', BinaryOperation.Operator.AND),
-        ('.or.', BinaryOperation.Operator.OR)])
+        ('.or.', BinaryOperation.Operator.OR),
+        ('sign', BinaryOperation.Operator.SIGN),
+        ('sum', BinaryOperation.Operator.SUM),
+        ('max', BinaryOperation.Operator.MAX),
+        ('min', BinaryOperation.Operator.MIN),
+        ('mod', BinaryOperation.Operator.MODULUS)])
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: Done.

File: src/psyclone/psyGen.py [hunk 5]
```diff
@@ -7334,21 +7356,37 @@ def _unary_op_handler(self, node, parent):
         Transforms an fparser2 UnaryOpBase to the PSyIR representation.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: Oops, well spotted.

File: src/psyclone/psyGen.py [hunk 6]
```diff
@@ -7357,31 +7395,145 @@ def _unary_op_handler(self, node, parent):
...
+        # to process the `items` of that...
+        self.process_nodes(parent=nary_op, nodes=list(node.items[1].items),
+                           nodes_parent=node.items[1])
+        return nary_op
+
+    def _intrinsic_handler(self, node, parent):
+        '''
+        Transforms an fparser2 Intrinsic_Function_Reference to the PSyIR
+        representation. Since Fortran Intrinsics can be unary, binary or
+        nary this handler identifies the appropriate 'sub handler'
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: Oh yes. Fixed.

File: src/psyclone/psyir/backend/fortran.py [hunk 7]
```diff
@@ -128,6 +128,27 @@ def gen_kind(symbol):
...
 
+def _reverse_map(op_map):
+    '''
+    Reverses the supplied fortran2psyir mapping to make a psyir2fortran
+    mapping.
+
+    :param op_map: mapping from string representation of operator to \
+                   enumerated type.
+    :type op_map: collections.OrderedDict
+
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/backend/fortran.py]: Done.

File: src/psyclone/psyir/backend/fortran.py [hunk 8]
```diff
@@ -264,31 +285,46 @@ def binaryoperation_node(self, node):
...
-                mapping[mapping_key] = mapping_value
+        mapping = _reverse_map(f2psyir.binary_operators)
         lhs = self._visit(node.children[0])
         rhs = self._visit(node.children[1])
         try:
             oper = mapping[node.operator]
             # This is a binary operation
+            if oper.upper() in FORTRAN_INTRINSICS:
+                # This is a binary intrinsic function.
+                return "{0}({1},{2})".format(oper.upper(), lhs, rhs)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/backend/fortran.py]: Yes. Done.

File: src/psyclone/psyir/backend/fortran.py [hunk 9]
```diff
@@ -264,31 +285,46 @@ def binaryoperation_node(self, node):
...
             return "{0}{1}{2}".format(lhs, oper, rhs)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/backend/fortran.py]: Done.

File: src/psyclone/psyir/backend/fortran.py [hunk 10]
```diff
@@ -264,31 +285,46 @@ def binaryoperation_node(self, node):
...
+        # Reverse the fortran2psyir mapping to make a psyir2fortran
+        # mapping.
+        from psyclone.psyGen import Fparser2ASTProcessor as f2psyir
+        mapping = _reverse_map(f2psyir.nary_operators)
+        arg_list = []
+        for child in node.children:
+            arg_list.append(self._visit(child))
+        try:
+            oper = mapping[node.operator]
+            return "{0}({1})".format(oper.upper(), ",".join(arg_list))
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/backend/fortran.py]: Yes, done.

File: src/psyclone/tests/psyGen_test.py [hunk 11]
```diff
@@ -4614,20 +4644,42 @@ def test_fparser2astprocessor_handling_part_ref(f2008_parser):
...
+        ('x = acos(a)', UnaryOperation, UnaryOperation.Operator.ACOS),
+        ('x = tan(a)', UnaryOperation, UnaryOperation.Operator.TAN),
+        ('x = atan(a)', UnaryOperation, UnaryOperation.Operator.ATAN),
+        ('x = real(a)', UnaryOperation, UnaryOperation.Operator.REAL),
+        ('x = real(a, 8)', CodeBlock, None),
+        ('x = int(a)', UnaryOperation, UnaryOperation.Operator.INT),
+        ('x = int(a, 8)', CodeBlock, None),
+        ('x = log(a)', UnaryOperation, UnaryOperation.Operator.LOG),
+        ('x = log10(a)', UnaryOperation, UnaryOperation.Operator.LOG10),
+        ('x = mod(a, b)', BinaryOperation, BinaryOperation.Operator.MODULUS),
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyGen_test.py]: Done.

**Comment by arporter** [comment]
All points addressed. Ready for next review.
(I think CodeCov is not reporting because it doesn't have the coverage of the point on master from which this branch was taken. I think that's because I was over-zealous in telling GitHub not to trigger Travis when doing the update of the changelog/top-level user guide in the final step of merging a branch to master. Lesson learned!)

File: src/psyclone/psyGen.py [hunk 12]
```diff
@@ -46,6 +46,7 @@
 import six
 from psyclone.configuration import Config
 from psyclone.core.access_type import AccessType
+from fparser.two import Fortran2003
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: Mine didn't but fixed :-)

File: src/psyclone/psyGen.py [hunk 13]
```diff
@@ -5938,67 +5933,48 @@ def gen_c_code(self, indent=0):
...
 
+    :raises TypeError: if the supplied operator is not an instance of \
+                       self.Operator.
+
+    '''
     def __init__(self, operator, parent=None):
-        super(UnaryOperation, self).__init__(parent=parent)
+        super(Operation, self).__init__(parent=parent)
 
         if not isinstance(operator, self.Operator):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: Good idea, done.

File: src/psyclone/psyGen.py [hunk 14]
```diff
@@ -5938,67 +5933,48 @@ def gen_c_code(self, indent=0):
...
-                       SCHEDULE_COLOUR_MAP["UnaryOperation"])
+        return self._operator
 
     def view(self, indent=0):
         '''
         Print a representation of this node in the schedule to stdout.
-
         :param int indent: level to which to indent output.
         '''
         print(self.indent(indent) + self.coloured_text + "[operator:'" +
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: Good point, done.

File: src/psyclone/psyGen.py [hunk 15]
```diff
@@ -6080,38 +6093,23 @@ def function_format(function_str, expr_str):
...
-class BinaryOperation(Node):
+class BinaryOperation(Operation):
     '''
     Node representing a BinaryOperation expression. As such it has two operands
     as children 0 and 1, and an attribute with the operator type.
 
-    :param operator: Enumerated type capturing the binary operator.
-    :type operator: :py:class:`psyclone.psyGen.BinaryOperation.Operator`
-    :param parent: the parent node of this BinaryOperation in the PSyIR.
-    :type parent: :py:class:`psyclone.psyGen.Node`
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: OK, done.

File: src/psyclone/psyGen.py [hunk 16]
```diff
@@ -6225,6 +6195,30 @@ def function_format(function_str, expr1, expr2):
                          self.children[1].gen_c_code())
 
 
+class NaryOperation(Operation):
+    '''
+    Node representing a n-ary operation expression. The n operands are the
+    stored as the 0 - n-1th children of this node and the type of the operator
+    is held in an attribute.
+
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: Done.

File: src/psyclone/psyGen.py [hunk 17]
```diff
@@ -6225,6 +6195,30 @@ def function_format(function_str, expr1, expr2):
...
+        :returns: Name of node + control chars for colour.
+        :rtype: str
+        '''
+        return colored("NaryOperation", SCHEDULE_COLOUR_MAP["Operation"])
+
+
 class Array(Reference):
     '''
     Node representing an Array reference. As such it has a reference and a
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: Done.

File: src/psyclone/tests/psyGen_test.py [hunk 18]
```diff
@@ -4637,18 +4689,130 @@ def test_fparser2astprocessor_handling_intrinsics(f2008_parser):
...
-        processor.process_nodes(fake_parent, [fp2node], None)
-    assert "Unexpected fparser2 node when parsing the real() intrinsic, 2 " \
-           "items were expected but found" in str(error.value)
+
+def test_fp2astproc_unary_op_handler_error(f2008_parser):
+    ''' Check that the unary op handler raises the expected error if the
+    parse tree has an unexpected structure. This is a hard error to
+    provoke since fparser checks that the number of arguments is correct. '''
+    from fparser.common.readfortran import FortranStringReader
+    from fparser.two.Fortran2003 import Execution_Part, Name
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyGen_test.py]: Bizarrely, `Execution_Part` *is* used but `Name` isn't so I removed that one :-)

File: src/psyclone/tests/psyGen_test.py [hunk 19]
```diff
@@ -4637,18 +4689,130 @@ def test_fparser2astprocessor_handling_intrinsics(f2008_parser):
...
+        processor._nary_op_handler(fp2node, fake_parent)
+    assert ("Expected second 'item' of N-ary intrinsic 'SUM(dummy)' in fparser"
+            " parse tree to be an Actual_Arg_Spec_List" in str(err))
+
+
+def test_fp2astproc_handling_nested_intrinsic(f2008_parser):
+    ''' Check that we correctly handle nested intrinsic functions. '''
+    from fparser.common.readfortran import FortranStringReader
+    from fparser.two.Fortran2003 import Execution_Part
+    from psyclone.psyGen import Reference
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyGen_test.py]: Removed.

File: src/psyclone/tests/psyGen_test.py [hunk 20]
```diff
@@ -4637,18 +4689,130 @@ def test_fparser2astprocessor_handling_intrinsics(f2008_parser):
...
+    assert "sum" not in [str(ref.name) for ref in array_refs]
+
+
+@pytest.mark.xfail(reason="#412 Fortran array notation not yet handled in "
+                   "non-NEMO PSyIR")
+def testfp2astproc_handling_array_product(f2008_parser):
+    ''' Check that we correctly handle array products. '''
+    from fparser.common.readfortran import FortranStringReader
+    from fparser.two.Fortran2003 import Execution_Part
+    from psyclone.psyGen import CodeBlock
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyGen_test.py]: Removed.

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 21]
```diff
@@ -317,14 +317,15 @@ def test_fw_binaryoperator():
...
-    assert "a=SIGN(1.0,1.0)" in result
+    assert "a=SIGN(1.0, 1.0)" in result
 
 
 def test_fw_binaryoperator_unknown(monkeypatch):
     '''Check the FortranWriter class binary_operation method raises an
     exception if an unknown binary operator is found.
 
     '''
+    from psyclone.psyGen import Fparser2ASTProcessor
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: Oops, apologies.

**Comment by arporter** [comment]
I've added some documentation and fixed all the other issues (apologies for missing the pylint stuff). As discussed, the auto-generated doc for the enumerations isn't great but I propose to live with it for the time being as we don't want to manually replicate the lists in the documentation.

**Pull request merged**